usr/sbin/pkcscca/pkcscca.c

```diff
@@ -1537,11 +1538,15 @@ int find_wrapped_keys(CK_FUNCTION_LIST *funcs, CK_SESSION_HANDLE sess,
         if (ulObjectCount == 0)
             break;
 
-        handles = realloc(handles, sizeof(CK_OBJECT_HANDLE) * (++ulTotalCount));
-        if (!handles) {
-            print_error("Malloc of %lu bytes failed!", ulTotalCount);
-            break;
+        ptr = realloc(handles, sizeof(CK_OBJECT_HANDLE) * (++ulTotalCount));
+        if (!ptr) {
+            print_error("Malloc of %lu bytes failed!",
+                        sizeof(CK_OBJECT_HANDLE) * ulTotalCount);
+            funcs->C_FindObjectsFinal(sess);
+            free(handles);
```

Comment:
hm, if realloc() fails, the original block is left untouched; it is not freed or moved.
Why do you then free handle ? the old code did the right thing here.

Reply:
If realloc fails it returns NULL and thus sets handles to NULL. The original block is not freed, but you don't have the address anymore, because handles was set to NULL. So the original block is leaked. 
The new code frees the original block when realloc fails, and cleas up and returns. The original code continued with handles being NULL and thus causes a null pointer dereference as well in later code.

Reply:
Would it be possible, to leave the handle memory untouched if realloc() fails and just
return with some failed. I agree, the old code didn't make it correct, but the fix now
could do better - and not free the handle memory but just return with failed.

Reply:
If you leave handles untouched when realloc fails and you break the loop (as it was before), then you would process only a subset of all the objects found. The user would not notice that you processed only a subset of the objects. Also ulTotalCount would have been incremented already, so it would actually not match the elements in handles, which causes another problem....

In my opinion: If realloc fails because we run out of memory (very unlikely anyway), it is very likely that any subsequent processing will also fail sooner or later, because it will  also allocate memory. So lets stop processing right now with a proper error message, and do proper cleanup.